### PR TITLE
feat: Introduce error interceptor

### PR DESCRIPTION
**Pull request by jhasan95**
### Handle error

- Exception Handling: Implemented errorInterceptor for non-2xx responses concerning status code.
- Error Handler Integration: Seamlessly added a robust error handler.

### CODE REVIEW

File: composeApp/src/commonMain/kotlin/data/network/ApiDefinition.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+sealed class ApiDefinition {
+    sealed class ApiEndpoint(val path: String) {
+        data object Login : ApiEndpoint(path = "auth/login")
+        data object Products : ApiEndpoint(path = "products")
+    }
+
+    object ApiField {
+
+        const val username = "username"
```
Comment (ShaQib07) [inline composeApp/src/commonMain/kotlin/data/network/ApiDefinition.kt]: use all uppercase letters while declaring constants.
const val USERNAME = "username"

File: composeApp/src/commonMain/kotlin/data/network/ApiErrorInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+             HttpStatusCode.NotFound -> throw ApiErrorException(statusCode, "Resource not found")
+             HttpStatusCode.InternalServerError -> throw ApiErrorException(statusCode, "Internal server error")
+             HttpStatusCode.ServiceUnavailable -> throw ApiErrorException(statusCode, "Service unavailable")
+             HttpStatusCode.RequestTimeout -> throw ApiErrorException(statusCode, "The request timed out")
+            else -> {
+                throw ApiErrorException(statusCode, "Unexpected status code: ${statusCode.value}")
+            }
+        }
+    }
+}
```
Comment (ShaQib07) [inline composeApp/src/commonMain/kotlin/data/network/ApiErrorInterceptor.kt]: Add new line

File: composeApp/src/commonMain/kotlin/data/network/ApiServiceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+class ApiServiceImpl(private val httpClient: HttpClient) : ApiService {
+    override suspend fun getProducts(limit: Int, skip: Int): HttpResponse = httpClient.get(
+        Products.path,
+    ) {
+        parameter(PARAM_LIMIT, limit)
+        parameter(PARAM_SKIP, skip)
+    }
+
+    override suspend fun authenticate(userName: String, pass: String): HttpResponse =
+        throw ApiErrorException(HttpStatusCode.BadRequest,"Bad ")
```
Comment (ShaQib07) [inline composeApp/src/commonMain/kotlin/data/network/ApiServiceImpl.kt]: Are you throwing this error for test purpose?

File: composeApp/src/commonMain/kotlin/presentation/feature/login/LoginContract.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    val password: String = "0lelplR",
+    val isLoading: Boolean = false,
+    val error: String? = null,
+    val loginButtonEnabled: Boolean = false,
+    val isLoggedIn: Boolean = false,
+)
+
+/**
+ * Login Actions emitted from the UI Layer
+ * passed to the coordinator to handle
```
Comment (ShaQib07) [inline composeApp/src/commonMain/kotlin/presentation/feature/login/LoginContract.kt]: Replace coordinator with viewmodel

**Review by ShaQib07 — APPROVED** [review]

File: composeApp/src/commonMain/kotlin/di/ApiErrorValidator.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package di
+
+import io.ktor.client.statement.HttpResponse
+
+interface ApiErrorValidator {
+    fun validateResponse(response: HttpResponse)
+}
```
Comment (levinzonr) [inline composeApp/src/commonMain/kotlin/di/ApiErrorValidator.kt]: DI package seems like not the best place to store something like this

File: composeApp/src/commonMain/kotlin/di/NetworkModule.kt [hunk 6]
```diff
@@ -17,17 +21,31 @@ import kotlinx.serialization.json.Json
...
     val networkClient = module {
         single {
             HttpClient {
+                expectSuccess = false
+                HttpResponseValidator {
+                    handleResponseExceptionWithRequest { exception, request ->
+                        val clientException = exception as? ClientRequestException
+                            ?: return@handleResponseExceptionWithRequest
+                        val exceptionResponse = clientException.response
+                        ApiErrorInterceptor().validateResponse(exceptionResponse)
```
Comment (levinzonr) [inline composeApp/src/commonMain/kotlin/di/NetworkModule.kt]: What's the point of and Interface if we just create an instance of a class extending it? Should we try injecting an interface instead?

File: composeApp/src/commonMain/kotlin/domain/usecase/LoginUseCase.kt [hunk 7]
```diff
@@ -14,8 +12,7 @@ class LoginUseCase(
         emit(
             runCatching {
                 authRepository.authenticate(email, password)
-            }
+            },
         )
     }
-        .flowOn(Dispatchers.Default)
```
Comment (levinzonr) [inline composeApp/src/commonMain/kotlin/domain/usecase/LoginUseCase.kt:20]: No point in this running as flow I think, can be just suspend function

File: composeApp/src/commonMain/kotlin/presentation/model/ErrorModel.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    data class ApiError(val exception: ApiErrorException, override val message: String) : ErrorModel
+
+    sealed class Connection : ErrorModel {
+        data object Timeout : Connection() {
+            override val message: String = "Connection timed out"
+        }
+    }
+
+    data class Unknown(val throwable: Throwable, override val message: String) : ErrorModel
+}
```
Comment (levinzonr) [inline composeApp/src/commonMain/kotlin/presentation/model/ErrorModel.kt]: can be sealed class with a message param in constructor

Comment (levinzonr) [reply composeApp/src/commonMain/kotlin/presentation/model/ErrorModel.kt]: Or keep an interface, but then lets maybe make `message` an extension property.

File: composeApp/src/commonMain/kotlin/data/network/errorhandling/ApiErrorInterceptor.kt [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                "Service unavailable",
+            )
+            HttpStatusCode.RequestTimeout -> throw ApiErrorException(
+                statusCode,
+                "The request timed out",
+            )
+            else -> {
+                throw ApiErrorException(statusCode, "Unexpected status code: ${statusCode.value}")
+            }
+        }
```
Comment (levinzonr) [inline composeApp/src/commonMain/kotlin/data/network/errorhandling/ApiErrorInterceptor.kt]: This when statement doesn't seem that necessary to me, all it does is checking the code and putting it right back to another class.  The message can be an extension function based on the status code and it should not be hardcoded (if it is meant for UI that is)

File: composeApp/src/commonMain/kotlin/data/network/ApiServiceImpl.kt [hunk 10]
```diff
@@ -23,6 +23,6 @@ class ApiServiceImpl(private val httpClient: HttpClient) : ApiService {
 
     override suspend fun authenticate(userName: String, pass: String): HttpResponse =
         httpClient.post(Login.path) {
-            setBody(mapOf(username to userName, password to pass))
+            setBody(mapOf(USERNAME to userName, PASSWORD to pass))
```
Comment (levinzonr) [inline composeApp/src/commonMain/kotlin/data/network/ApiServiceImpl.kt]: Why not create a dedicated DTO for this and use Kotlin serialization?

**Review by levinzonr — CHANGES_REQUESTED** [review]
Left some comments :)
When it comes to the error handling I would prefer to it to be closer to what we have in the Android Template now
```
sealed interface ErrorModel {

    data class ApiError(val exception: ApiException) : ErrorModel

    sealed class Connection : ErrorModel {
        object Timeout : Connection()
        object IOError : Connection()
        object UnknownHost : Connection()
    }

    data class Unknown(val throwable: Throwable) : ErrorModel
}
```

File: composeApp/src/commonMain/kotlin/data/network/errorhandling/ApiErrorInterceptor.kt (hunk 9 above)
Comment (levinzonr) [reply composeApp/src/commonMain/kotlin/data/network/errorhandling/ApiErrorInterceptor.kt]: I think what we really should do here is to try and convert non 200 responses and their response bodies into domain error exception. I.e something like that
```data class ApiException(
    val code: Int,
    val displayableMessage: String,
    val loggableMessage: String
) : IOException(displayableMessage)
```

File: composeApp/src/commonMain/kotlin/domain/usecase/LoginUseCase.kt (hunk 7 above)
Comment (jhasan95) [reply composeApp/src/commonMain/kotlin/domain/usecase/LoginUseCase.kt:20]: Flow provides lifecycle-aware operations like **onStart** and **onCompletion**, allowing for actions at the beginning and end of the flow's lifecycle, such as displaying and concealing progress indicators.

Comment (levinzonr) [reply composeApp/src/commonMain/kotlin/domain/usecase/LoginUseCase.kt:20]: you can do the same in the beginning and in the end of the suspend call. I feel like this complicates things in various of ways. 
- The flow is very powerful tool, and using its APIs for just one emit call is an overkill
- It doesn't make the code more readable, onCompleteion in this scenario is the same as collect { }

**Review by levinzonr — APPROVED** [review]

**Pull request merged**